xla/service/gpu/cudnn_fused_mha_rewriter.cc

```diff
@@ -1393,67 +1546,72 @@ StatusOr<bool> FuseBwdMultiHeadedAttentionBlock(
       lhs_bmm1_grad_gemm2,
       ChangeCheckedDimToFastest(comp, bmm_1_grad_2, false /*is_lhs*/,
                                 false /*should_contracting_be_fastest*/));
-  // Forward activation
+  // P tensor
   TF_ASSIGN_OR_RETURN(
       lhs_bmm2_grad_gemm1,
       ChangeCheckedDimToFastest(comp, bmm_2_grad_1, true /*is_lhs*/,
                                 false /*should_contracting_be_fastest*/));
+
+  // Forward activation
+  // if it is not flash attention, fwd activation is the P tensor
+  // else it is the softmax_stats
+  if (fwd_config.is_flash_attention()) {
+    auto fwd_act_index = 2;
+    fwd_act = comp->AddInstruction(HloInstruction::CreateGetTupleElement(
+        fwd_fmha_call->shape().tuple_shapes(fwd_act_index), fwd_fmha_call,
+        fwd_act_index));
+  } else {
+    fwd_act = lhs_bmm2_grad_gemm1;
+  }
+
   // V tensor
   TF_ASSIGN_OR_RETURN(
       rhs_bmm2_grad_gemm2,
       ChangeCheckedDimToFastest(comp, bmm_2_grad_2, false /*is_lhs*/,
                                 true /*should_contracting_be_fastest*/));
-  // d output
+  // d output to bmm2_grad2
   // Since d_o is the input of 2 bmms, we set the dim number using the
   // constraint
   // -> the contracting dimension of the lhs of bmm_2_grad_2 needs to be the
   // fastest moving dimension.
-  TF_ASSIGN_OR_RETURN(d_output_grad, ChangeCheckedDimToFastest(
-                                         comp, bmm_2_grad_2, true /*is_lhs*/,
-                                         true /*check_contracting_dim*/));
-  // Operand order {Q, K, V, Fwd act, d_o, mask*}
+  TF_ASSIGN_OR_RETURN(
+      d_output_grad,
+      ChangeCheckedDimToFastest(comp, bmm_2_grad_2, true /*is_lhs*/,
+                                true /*should_contracting_be_fastest*/));
+  // d output to bmm2_grad1
+  // we don't use this value but we call this to make sure dot number is being
+  // set correctly
+  TF_ASSIGN_OR_RETURN(
+      HloInstruction * bmm_2_grad_1_rhs,
+      ChangeCheckedDimToFastest(comp, bmm_2_grad_1, false /*is_lhs*/,
+                                false /*should_contracting_be_fastest*/));
```

Comment:
Can you please add

(void)bmm_2_grad_1_rhs;

so that there will be no unused variable warning?